substrate-node/pallets/pallet-tfgrid/src/lib.rs

```diff
@@ -2327,6 +2368,70 @@ impl<T: Config> Pallet<T> {
         let parsed_serial_number = <T as Config>::SerialNumber::try_from(serial_number)?;
         Ok(parsed_serial_number)
     }
+
+    fn _change_power_state(
+        account_id: T::AccountId,
+        power_state: Power,
+    ) -> DispatchResultWithPostInfo {
+        let twin_id = TwinIdByAccountID::<T>::get(&account_id).ok_or(Error::<T>::TwinNotExists)?;
+        ensure!(
+            NodeIdByTwinID::<T>::contains_key(twin_id),
+            Error::<T>::NodeNotExists
+        );
+        let node_id = NodeIdByTwinID::<T>::get(twin_id);
+        let node = Nodes::<T>::get(node_id).ok_or(Error::<T>::NodeNotExists)?;
+
+        let power_state = match power_state {
+            Power::Up => PowerState::Up,
+            Power::Down => PowerState::Down(system::Pallet::<T>::block_number()),
```

Comment:
Why do we need the block number? Does ZOS need it? Don't we know it from the event?

Reply:
We save the blocknumber on this powerstate for the minting to check later